src/languages/rust.ts

```diff
@@ -0,0 +1,175 @@
+ * @param node The node which we might match
+ * @returns The return value node
+ */
+function returnValueFinder(node: SyntaxNode) {
+  if (node.type !== "block") {
+    return null;
+  }
+
+  const { lastNamedChild } = node;
+
+  // The return expression will always be the last statement or expression in
+  // the block
+  if (lastNamedChild == null) {
+    return null;
+  }
+
+  // If the final name child is an expression statement not a raw expression
+  // then we only treat it as a return value if it is a return expression.
+  // Otherwise it is just a normal statement that doesn't return anything
+  if (lastNamedChild.type === "expression_statement") {
+    const expression = lastNamedChild.child(0)!;
+
+    if (expression.type === "return_expression") {
+      return expression.child(1);
+    }
+
+    return null;
+  }
+
+  // Any other type of statement is not a return statement so we should not
+  // match it
+  if (STATEMENT_TYPES.includes(lastNamedChild.type)) {
+    return null;
+  }
+
+  // NB: At this point we have now excluded all statement types of the only other
+  // possible node is an expression node
+
+  // If it is a return expression then we zoom down to the actual value of the
+  // return expression. This happens when they say `return foo` with no
+  // trailing semicolon
+  if (lastNamedChild.type === "return_expression") {
+    return lastNamedChild.child(1);
+  }
+
+  // At this point it is an expression which is not a return expression so we
+  // just return it as the return value of the block
+  return lastNamedChild;
+}
+
+const nodeMatchers: Partial<
+  Record<SimpleScopeTypeType, NodeMatcherAlternative>
+> = {
+  statement: ancestorChainNodeMatcher(
+    [
+      patternFinder(...STATEMENT_PARENT_TYPES),
+      patternFinder(...STATEMENT_TYPES),
+    ],
+    1
+  ),
+  string: ["raw_string_literal", "string_literal"],
+  ifStatement: ["if_expression", "if_let_expression"],
+  functionCall: ["call_expression", "macro_invocation", "struct_expression"],
+  functionCallee: "call_expression[function]",
+  comment: ["line_comment", "block_comment"],
+  list: ["array_expression", "tuple_expression"],
+  collectionItem: argumentMatcher("array_expression", "tuple_expression"),
+  namedFunction: "function_item",
+  type: cascadingMatcher(
+    leadingMatcher(
+      ["let_declaration[type]", "parameter[type]", "field_declaration[type]"],
+      [":"]
+    ),
+    leadingMatcher(["function_item[return_type]"], ["->"]),
+    matcher(traitTypeMatcher),
+    patternMatcher("struct_item", "trait_item", "impl_item")
+  ),
+  functionName: ["function_item[name]"],
+  anonymousFunction: "closure_expression",
+  argumentOrParameter: argumentMatcher("arguments", "parameters"),
+  name: [
+    "let_declaration.identifier!",
+    "parameter.identifier!",
+    "function_item[name]",
+    "struct_item[name]",
+    "enum_item[name]",
+    "trait_item[name]",
+    "const_item[name]",
```

Comment:
you could maybe just replace all of these by `*[name]`.  Are you trying to exclude something in particular?

Reply:
Not really, but I don’t want to match something unintentionally now or in the future.

Reply:
We do this in lots of places; remember that the cost of matching something that you don't want is low: you just end up stopping a little bit lower in the parse tree than you hoped.  And it's quite unlikely that if it's called "name", it's not what you want.  I'd actually argue that you're future-proofing yourself by doing so